chain/client/src/sync/state.rs

```diff
@@ -740,8 +616,27 @@ impl StateSync {
                         state_parts_future_spawner,
                         self.state_parts_mpsc_tx.clone(),
                     );
-                    if semaphore.available_permits() == 0 {
-                        break;
+                }
+            } else {
+                if peer_requests_sent >= MAX_STATE_PART_REQUEST {
+                    continue;
+                }
+
+                // The request sent to the network adapater needs to include the sync_prev_prev_hash
+                // so that a peer hosting the correct snapshot can be selected.
+                if let Ok(header) = chain.get_block_header(&sync_hash) {
+                    if let Ok(prev_header) = chain.get_block_header(&header.prev_hash()) {
+                        let sync_prev_prev_hash = prev_header.prev_hash();
+                        request_part_from_peers(
```

Comment:
Yeah interesting I'm able to reproduce what you're describing using the test and diff that you shared. The chain sometimes (but not always) stalls after producing block 32. This occurs exactly when the nodes fail to advertise a snapshot at epoch height 2; I am still working on debugging why that can happen.

I do also see the problem you are describing that the snapshots are always claiming to have all shards. We just publish whatever this function returns:

https://github.com/near/nearcore/blob/d6d6e4782f7d57a7b7acc2894e1668355012ff74/core/store/src/trie/state_snapshot.rs#L171-L172

to the network here:

https://github.com/near/nearcore/blob/d6d6e4782f7d57a7b7acc2894e1668355012ff74/chain/chain/src/state_snapshot_actor.rs#L95

so it seems like it might be a bug inside the snapshot creation side of things.

Reply:
Regarding the timeout, it is explicitly set to 2 seconds in this test ([see here](https://github.com/near/nearcore/blob/d6d6e4782f7d57a7b7acc2894e1668355012ff74/pytest/tests/sanity/transactions.py#L30-L72)) but we use [60 seconds in mainnet](https://github.com/near/nearcore/blob/d6d6e4782f7d57a7b7acc2894e1668355012ff74/core/chain-configs/src/client_config.rs#L281-L283).

Reply:
Found and fixed the reason that specifically the second snapshot wasn't getting advertised properly:

https://github.com/near/nearcore/pull/12111/files#diff-b40c686ecfc158bc56c26a76881529c88ad148956cb668453fa9a00da8660954R941-R949

Reply:
Added [this commit ](https://github.com/near/nearcore/pull/12111/commits/cbd28985be3ad23b060b0cf694230f5586bf6162) which makes it so that we pass the tracked shards to the snapshot actor instead of indiscriminately requesting a snapshot of all shards. It has the downstream effect that SnapshotHostInfos are accurate now with respect to tracked shards.

Based on some testing with these changes I believe everything is working well. The decentralized sync seems to be a bit too slow for 10-block epochs ([one culprit for latency is here for example](https://github.com/near/nearcore/pull/12110/files#diff-b40c686ecfc158bc56c26a76881529c88ad148956cb668453fa9a00da8660954R92)), but if I set epoch length to 20 blocks I can observe the nodes shuffling their shard assignment, syncing from each other, and keeping up with the chain.

<img width="120" alt="image" src="https://github.com/user-attachments/assets/0dd40690-0b0e-4861-ab0b-60ad2acebb6a">

Reply:
Upon further investigation, I believe there is a bug in the snapshot creation. If you apply only cbd28985be3ad23b060b0cf694230f5586bf6162 to master branch, it breaks some tests because a bad snapshot is produced when not all shards are requested. Subsequent calls to generate state parts from the snapshot fail.

For now, we can use the fix in #12143 instead to ensure that nodes which do not track all shards publish the correct SnapshotHostInfo.